### PR TITLE
Save channels in datagrid

### PR DESCRIPTION
**Pull request by dominik-zeglen**
I want to merge this change because adds ability to edit pricing and availability of variants in the context of channels.

**PR intended to be tested with API branch:** main

### Pull Request Checklist

1. [ ] This code contains UI changes
2. [ ] All visible strings are translated with proper context including data-formatting
3. [ ] Attributes `[data-test-id]` are added for new elements
4. [ ] Changes are mentioned in the changelog
5. [ ] The changes are tested in different browsers and in light/dark mode

### Test environment config

API_URI=https://automation-dashboard.staging.saleor.cloud/graphql/
MARKETPLACE_URL=https://marketplace-gray.vercel.app/

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-7560 Add channel editing and saving](https://app.clickup.com/t/2549495/SALEOR-7560)

**Comment by patrys** [comment]
Task linked: [SALEOR-6661 Datagrid](https://app.clickup.com/t/2549495/SALEOR-6661)

File: src/components/Datagrid/Datagrid.stories.tsx [hunk 1]
```diff
@@ -42,58 +43,40 @@ const DefaultStory: React.FC = () => {
...
 
       if (columnId === "loan-active") {
-        const value = change ?? initialData[row].loan.active;
         return {
-          ...baseProps,
-          allowOverlay: false,
-          kind: GridCellKind.Boolean,
-          data: value,
+          ...errorProps,
+          ...booleanCell(change ?? dataRow?.loan.active ?? null),
```
Comment (Droniu) [inline src/components/Datagrid/Datagrid.stories.tsx:48]:
```suggestion
          ...booleanCell(change ?? (dataRow?.loan.active ?? null)),
```

File: src/components/Datagrid/Datagrid.stories.tsx [hunk 2]
```diff
@@ -42,58 +43,40 @@ const DefaultStory: React.FC = () => {
...
 
       if (columnId === "balance") {
-        const data = change ?? initialData[row][columnId];
         return {
-          ...baseProps,
-          kind: GridCellKind.Number,
-          data,
-          displayData: data?.toString(),
+          ...errorProps,
+          ...numberCell(change ?? dataRow?.balance ?? numberCellEmptyValue),
```
Comment (Droniu) [inline src/components/Datagrid/Datagrid.stories.tsx:65]:
```suggestion
          ...numberCell(change ?? (dataRow?.balance ?? numberCellEmptyValue)),
```

File: src/components/Datagrid/cells.ts [hunk 3]
```diff
@@ -17,19 +20,30 @@ export function textCell(value: string): GridCell {
   };
 }
 
-export function numberCell(value: number | null): NumberCell {
+export function booleanCell(value: boolean): GridCell {
```
Comment (Droniu) [inline src/components/Datagrid/cells.ts:23]: Why can other cells take nullish values whereas boolean does not?

File: src/components/Datagrid/styles.ts [hunk 4]
```diff
@@ -67,4 +78,29 @@ const useStyles = makeStyles(
...
+      accentColor: theme.palette.saleor.main[1],
+      accentLight: theme.palette.background.paper,
+      accentFg: theme.palette.divider,
+      bgCell: theme.palette.background.default,
+      bgHeader: theme.palette.background.default,
+      bgHeaderHasFocus: theme.palette.background.default,
+      bgHeaderHovered: theme.palette.background.default,
+      bgBubbleSelected: theme.palette.background.paper,
+      textHeader: theme.palette.saleor.main[3],
+      borderColor: "rgba(37, 41, 41, 0.1)",
```
Comment (Droniu) [inline src/components/Datagrid/styles.ts:94]: 👀

File: src/graphql/client.ts [hunk 5]
```diff
@@ -1,18 +1,31 @@
...
+  })),
+);
+
+export const link = attachVariablesLink.concat(
+  createUploadLink({
+    credentials: "include",
+    uri: API_URI,
+    fetch: createFetch(),
+  }),
+);
```
Comment (Droniu) [inline src/graphql/client.ts:28]: Why is this change needed?

File: src/products/components/ProductUpdatePage/ProductChannelsListingsDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  data,
+  open,
+  onClose,
+  onConfirm,
+}) => {
+  const intl = useIntl();
+
+  const [selected, setSelected] = useStateFromProps(
+    data.channels.updateChannels.map(listing => listing.channelId),
+  );
```
Comment (Droniu) [inline src/products/components/ProductUpdatePage/ProductChannelsListingsDialog.tsx:33]: I think we shouldn't use useStateFromProps in new components, it should be depreciated in favor of useEffect which does reference comparisons together with memoized initial value

File: src/products/components/ProductUpdatePage/formChannels.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  ProductFragment,
+} from "@saleor/graphql";
+import useStateFromProps from "@saleor/hooks/useStateFromProps";
+import uniq from "lodash/uniq";
+import uniqBy from "lodash/uniqBy";
+import { useCallback, useRef } from "react";
+
+import { ProductChannelsListingDialogSubmit } from "./ProductChannelsListingsDialog";
+
+const emptyListing = {
```
Comment (Droniu) [inline src/products/components/ProductUpdatePage/formChannels.ts:13]: please add type here

File: src/products/components/ProductUpdatePage/formChannels.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  isPublished: false,
+  publicationDate: null,
+  visibleInListings: false,
+};
+
+export function useProductChannelListingsForm(
+  product: Pick<ProductFragment, "channelListings">,
+  triggerChange: () => void,
+) {
+  const [channels, setChannels] = useStateFromProps<
```
Comment (Droniu) [inline src/products/components/ProductUpdatePage/formChannels.ts:25]: another useStateFromProps with complex objects to deep compare

File: src/products/components/ProductVariants/utils.tsx [hunk 9]
```diff
@@ -120,69 +200,111 @@ interface GetData {
...
+    err => err.variantId === variantId && errorMatchesColumn(err, columnId),
+  );
+
+  const styled = (props: GridCell) => ({
+    ...props,
+    themeOverride: error
+      ? { bgCell: "#F4DDBA" }
+      : change !== undefined
+      ? {
+          bgCell: "#C1DBFF",
```
Comment (Droniu) [inline src/products/components/ProductVariants/utils.tsx:233]: Where do these fancy colors come from

File: src/products/components/ProductVariants/utils.tsx [hunk 10]
```diff
@@ -120,69 +200,111 @@ interface GetData {
...
+        readonly: true,
+        allowOverlay: false,
+      });
+    }
 
-  //   return moneyCell(value, currency);
-  // }
+    const currency = channels.find(channel => channelId === channel.id)
+      ?.currency;
+    const value = change?.value ?? listing?.price?.amount ?? 0;
```
Comment (Droniu) [inline src/products/components/ProductVariants/utils.tsx:275]:
```suggestion
    const value = change?.value ?? (listing?.price?.amount ?? 0);
```

File: src/products/components/ProductVariants/utils.tsx [hunk 11]
```diff
@@ -199,32 +321,42 @@ export function getColumnData(
...
+  }
+
+  if (getColumnChannel(name)) {
+    const channel = channels.find(
+      channel => channel.id === getColumnChannel(name),
+    );
+    return {
+      ...common,
+      width: 150,
+      title: `Price in ${channel.name} [${channel.currency}]`,
```
Comment (Droniu) [inline src/products/components/ProductVariants/utils.tsx:340]: This looks like should be translated

File: src/products/components/ProductVariants/utils.tsx [hunk 12]
```diff
@@ -199,32 +321,42 @@ export function getColumnData(
...
+  }
+
+  if (getColumnChannelAvailability(name)) {
+    const channel = channels.find(
+      channel => channel.id === getColumnChannelAvailability(name),
+    );
+    return {
+      ...common,
+      width: 80,
+      title: `Available in ${channel.name}`,
```
Comment (Droniu) [inline src/products/components/ProductVariants/utils.tsx:351]: Same here

File: src/products/components/ProductVariants/utils.tsx (hunk 11 above)
Comment (dominik-zeglen) [reply src/products/components/ProductVariants/utils.tsx:340]: It's going to be changed

File: src/products/components/ProductVariants/utils.tsx (hunk 12 above)
Comment (dominik-zeglen) [reply src/products/components/ProductVariants/utils.tsx:351]: This too, left it here before i get to styling

File: src/components/Datagrid/styles.ts [hunk 13]
```diff
@@ -33,9 +42,11 @@ const useStyles = makeStyles(
         },
       },
       portal: { position: "fixed", top: 0, left: 0 },
+      datagrid: {
+        // borderTop: `1px solid ${theme.palette.divider}`,
+        // borderBottom: `1px solid ${theme.palette.divider}`,
```
Comment (orzechdev) [inline src/components/Datagrid/styles.ts:47]: 👀

File: src/products/components/ProductVariants/utils.tsx [hunk 14]
```diff
@@ -120,69 +200,111 @@ interface GetData {
...
-  // if (isChannelColumn.test(columnId)) {
-  //   const channelId = columnId.match(isChannelColumn)[1];
-  //   const listing = dataRow?.channelListings.find(
-  //     listing => listing.channel.id === channelId,
-  //   );
-  //   const currency = channels.find(channel => channelId === channel.id)
-  //     ?.currencyCode;
-  //   const value = change?.value ?? listing?.price?.amount ?? 0;
+  if (getColumnChannel(columnId)) {
+    const channelId = getColumnChannel(columnId);
```
Comment (orzechdev) [inline src/products/components/ProductVariants/utils.tsx:257]:
```suggestion
const channelId = getColumnChannel(columnId);
  if (channelId) {
```

**Review by orzechdev — APPROVED** [review]

File: src/components/Datagrid/cells.ts (hunk 3 above)
Comment (dominik-zeglen) [reply src/components/Datagrid/cells.ts:23]: You can't clear this field, it's just true or false

File: src/components/Datagrid/styles.ts (hunk 4 above)
Comment (dominik-zeglen) [reply src/components/Datagrid/styles.ts:94]: Will be fixed in #2315

File: src/graphql/client.ts (hunk 5 above)
Comment (dominik-zeglen) [reply src/graphql/client.ts:28]: We need access not only to response but to variables too - since request count is not constant and queries don't have identifiers we need another way to map responses to changes in datagrid.

File: src/products/components/ProductUpdatePage/ProductChannelsListingsDialog.tsx (hunk 6 above)
Comment (dominik-zeglen) [reply src/products/components/ProductUpdatePage/ProductChannelsListingsDialog.tsx:33]: I know, but didn't want to go into rabbit hole here (as my time here is limited), so I just reused old code

File: src/products/components/ProductUpdatePage/formChannels.ts (hunk 8 above)
Comment (dominik-zeglen) [reply src/products/components/ProductUpdatePage/formChannels.ts:25]: I won't make refactors here as this feature is huge by itself

**Review by Droniu — APPROVED** [review]
